packages/cosmic-swingset/src/chain-main.js

```diff
@@ -401,6 +408,65 @@ export default async function main(progname, args, { env, homedir, agcc }) {
     const keepSnapshots =
       XSNAP_KEEP_SNAPSHOTS === '1' || XSNAP_KEEP_SNAPSHOTS === 'true';
 
+    const nodeHeapSnapshots = Number.parseInt(NODE_HEAP_SNAPSHOTS, 10);
+
+    let lastSnapshotTime = nodeHeapSnapshots >= 0 ? 0 : NaN;
+    let commitCallsSinceLastSnapshot = nodeHeapSnapshots > 0 ? -1 : NaN;
+    const snapshotBaseDir = path.resolve(stateDBDir, 'node-heap-snapshots');
+
+    let heapSnapshot;
+
+    if (nodeHeapSnapshots >= 0) {
+      fs.mkdirSync(snapshotBaseDir, { recursive: true });
+    }
+
+    const getRuntimeStats = () => {
+      const t0 = performance.now();
+      const heapStats = v8.getHeapStatistics();
+      const t1 = performance.now();
+      const heapSpaceStats = v8.getHeapSpaceStatistics();
+      const t2 = performance.now();
+      const memoryUsage = process.memoryUsage();
+      const t3 = performance.now();
+
+      commitCallsSinceLastSnapshot += 1;
+      if (
+        t0 - lastSnapshotTime > 30 * 1000 ||
+        (nodeHeapSnapshots > 0 &&
+          commitCallsSinceLastSnapshot >= nodeHeapSnapshots)
+      ) {
+        commitCallsSinceLastSnapshot -= commitCallsSinceLastSnapshot;
+        lastSnapshotTime = t0;
+        heapSnapshot = `Heap-${process.pid}-${Date.now()}.heapsnapshot`;
+      } else {
+        heapSnapshot = undefined;
+      }
+
+      return {
+        heapStats,
+        heapSpaceStats,
+        memoryUsage,
+        heapSnapshot,
+        statsTime: {
+          heapStats: t1 - t0,
+          heapSpaceStats: t2 - t1,
+          memoryUsage: t3 - t2,
+        },
```

Comment:
This is so we can check that gathering these memory stats does not significantly delay the commit. I originally wanted to measure the commit time itself, but it didn't feel right to plumb `performance` into `launch-chain`

Reply:
We now measure the commit time by emitting both a start and a finish for the commit-block event.